### PR TITLE
Fix incorrect panel height calculation in complex layout

### LINKED ISSUE #1390: Panel in layout adds newline in more complex layout split
**Information**
 - OS: MacOS
 - Version: 0.48.0, 0.47.0
 - Terminal: Kitty

**Describe the bug**
There's some issue with adding newline or setting the wrong height on panel. It occurs mostly when there are 3 rows in layout and height of AnsiConsole is a specific number e.g. 17, 20, 23, 28, 31. Changing console height to other "correct" number seems to resolve the problem, so I would assume it's correlated to some measurement calculation?

**To Reproduce**
Run this code and set height of terminal to specific number e.g. provided above or try to change height by one unit at a time, until issue occurs.

```csharp
using Spectre.Console;

var layout = new Layout();
var col1 = new Layout();
var col1Row1 = new Layout();
var col1Row2 = new Layout();
var col1Row3 = new Layout();

col1.SplitRows(col1Row1, col1Row2, col1Row3);

var col2 = new Layout();

layout.SplitColumns(col1, col2);

var panel = new Panel("Hello, World!") { Expand = true };

List<Layout> layouts = [col1Row1, col1Row2, col1Row3, col2];
foreach (var l in layouts)
{
    l.Update(panel);
}

AnsiConsole.Clear();
AnsiConsole.Write(layout);
Console.ReadKey();
```

**Expected behavior**
Display correct panels, without newline.

**Screenshots**
Should be:
image
but is:
image

---
Please upvote :+1: this issue if you are interested in it.

### PR DESCRIPTION
**Pull request by BlazeFace**
fixes #1390 

- [x] I have read the [Contribution Guidelines](https://github.com/spectreconsole/spectre.console/blob/main/CONTRIBUTING.md)
- [x] I have commented on the issue above and discussed the intended changes
- [ ] A maintainer has signed off on the changes and the issue was assigned to me
- [x] All newly added code is adequately covered by tests
- [x] All existing tests are still running without errors
- [x] The documentation was modified to reflect the changes _OR_ no documentation changes are required.

## Changes

Updated the ratio calculation to round up if the remainder is within 0.9999 I have tested this with a good amount of examples and it still works, but I have not written any proof that this will remain correct at any value. We could use a much smaller target to round say .99 as well.

---
Please upvote :+1: this pull request if you are interested in it.

### CODE REVIEW

**Comment by BlazeFace** [comment]
@patriksvensson The only major modification I was considering was a bounds check to see if the total height could go over the target, do you think that is needed?

**Comment by FrankRay78** [comment]
I'll review this @BlazeFace

**Comment by BlazeFace** [comment]
@FrankRay78 if we are looking to merge this I can go ahead and rebase and resolve conflicts.

**Comment by BlazeFace** [comment]
@FrankRay78 Rebase complete

File: src/Tests/Spectre.Console.Tests/Unit/Widgets/LayoutTests.cs [hunk 1]
```diff
@@ -264,4 +264,37 @@ public Task Should_Fall_Back_To_Parent_Layout_If_All_Children_Are_Invisible()
         // Then
         return Verifier.Verify(console.Output);
     }
+
+    [Fact]
```
Comment (FrankRay78) [inline src/Tests/Spectre.Console.Tests/Unit/Widgets/LayoutTests.cs:268]: Should we make this a [Theory] and pass in the problematic console heights 17, 20, 23, 28, 31 ?

File: src/Tests/Spectre.Console.Tests/Unit/Widgets/LayoutTests.cs [hunk 2]
```diff
@@ -264,4 +264,37 @@ public Task Should_Fall_Back_To_Parent_Layout_If_All_Children_Are_Invisible()
...
+    [Expectation("Render_Layout_With_Three_And_One_Columns")]
```
Comment (FrankRay78) [inline src/Tests/Spectre.Console.Tests/Unit/Widgets/LayoutTests.cs:269]: I think this/the underlying expectation file should probably be renamed to `Render_Layout_With_Nested_Three_Rows_In_One_Column`, to be more consistent with:

image

File: src/Tests/Spectre.Console.Tests/Unit/Widgets/LayoutTests.cs [hunk 3]
```diff
@@ -264,4 +264,37 @@ public Task Should_Fall_Back_To_Parent_Layout_If_All_Children_Are_Invisible()
...
+    public Task Should_Render_Layout_With_Three_And_One_Columns()
```
Comment (FrankRay78) [inline src/Tests/Spectre.Console.Tests/Unit/Widgets/LayoutTests.cs:270]: Could you please move this to line 156, so your test follows directly after `public Task Should_Render_Layout_With_Nested_Rows_And_Columns()`, which is a natural grouping.

File: src/Tests/Spectre.Console.Tests/Unit/Widgets/LayoutTests.cs [hunk 4]
```diff
@@ -264,4 +264,37 @@ public Task Should_Fall_Back_To_Parent_Layout_If_All_Children_Are_Invisible()
...
+
+        layout.SplitColumns(col1, col2);
+
+        var panel = new Panel("Hello, World!") { Expand = true };
+
+        List<Layout> layouts = [col1Row1, col1Row2, col1Row3, col2];
+        foreach (var l in layouts)
+        {
+            l.Update(panel);
+        }
```
Comment (FrankRay78) [inline src/Tests/Spectre.Console.Tests/Unit/Widgets/LayoutTests.cs:292]: I've thought about this, and whilst I'm happy with the correctness of the above test, all the other tests in the same file use a fluent convention. Can you please have a go at updating your test to be in the same style as the others, here's a close example to follow (I'm sorry for this 🏈 ache request, and I know the test comes direct from the issue repro code):

```csharp
    public Task Should_Render_Layout_With_Nested_Rows_And_Columns()
    {
        // Given
        var console = new TestConsole().Size(new Size(40, 15));
        var layout = new Layout()
            .SplitRows(
                new Layout("Top")
                    .SplitRows(
                        new Layout("T1")
                            .SplitColumns(
                                new Layout("A"),
                                new Layout("B")),
                        new Layout("T2")
                            .SplitColumns(
                                new Layout("C"),
                                new Layout("D"))),
                new Layout("Bottom")
                    .SplitRows(
                        new Layout("B1")
                            .SplitColumns(
                                new Layout("E"),
                                new Layout("F")),
                        new Layout("B2")
                            .SplitColumns(
                                    new Layout("G"),
                                    new Layout("H"))));

        // When
        console.Write(layout);

        // Then
        return Verifier.Verify(console.Output);
    }
```

**Review by FrankRay78 — CHANGES_REQUESTED** [review]
Hello @BlazeFace, if you'll be kind enough to entertain my review feedback, I'll be most kind enough to merge the PR once done. (So you know your efforts won't languish in an unmerged PR).

File: src/Tests/Spectre.Console.Tests/Unit/Widgets/LayoutTests.cs (hunk 1 above)
Comment (BlazeFace) [reply src/Tests/Spectre.Console.Tests/Unit/Widgets/LayoutTests.cs]: Done!

File: src/Tests/Spectre.Console.Tests/Unit/Widgets/LayoutTests.cs (hunk 2 above)
Comment (BlazeFace) [reply src/Tests/Spectre.Console.Tests/Unit/Widgets/LayoutTests.cs]: Done!

File: src/Tests/Spectre.Console.Tests/Unit/Widgets/LayoutTests.cs (hunk 3 above)
Comment (BlazeFace) [reply src/Tests/Spectre.Console.Tests/Unit/Widgets/LayoutTests.cs]: Done!

File: src/Tests/Spectre.Console.Tests/Unit/Widgets/LayoutTests.cs (hunk 4 above)
Comment (BlazeFace) [reply src/Tests/Spectre.Console.Tests/Unit/Widgets/LayoutTests.cs]: Done!

**Comment by FrankRay78** [comment]
Thanks @BlazeFace, I've seen lots of activity but not had the time to look just yet. But I will do so soon.

**Review by patriksvensson — APPROVED** [review]
LGTM 👍

**Comment by FrankRay78** [comment]
Good stuff @BlazeFace. Don't hesitate to tag me when you have other PRs to review.

**Comment by patriksvensson** [comment]
@FrankRay78 I think you forgot to approve and merge 😉

**Comment by FrankRay78** [comment]
A teacher knows to be humble before the master.

**Comment by patriksvensson** [comment]
LGTM 👍

**Pull request merged**